File: python/stack.py
from linked_list import LinkedList, Node

class Stack:
    def __init__(self) -> None:
        self.top = None
        self.length = 0
        self.data = []

    def peek(self):
        return self.top
    
    def push(self, value):
        self.data.append(value)
        self.top = value
        self.length += 1

    def pop(self):
        if self.length == 0:
            return None
        removed = self.data.pop(-1)
        self.length -= 1
        if self.length > 0:
            self.top = self.data[-1]
        else:
            self.top = None
        return removed
    
    def is_empty(self):
        if self.length != 0:
            return False
        return True
    

class Stack2:
    def __init__(self) -> None:
        self.top = None
        self.data = LinkedList()
        self.length = 0

    def peek(self):
        return self.top
    
    def push(self, value):
        new_node = Node(value)
        self.data.append(new_node)
        self.length += 1
        self.top = value

    def pop(self):
        if self.length == 0:
            return None
        length_data = self.data.length
        removed = self.data.delete(length_data-1)
        self.length -= 1
        if self.length > 0:
            self.top = self.data.get_item(self.data.length - 1)
        else:
            self.top = None  
        return removed


if __name__ == "__main__":
    stack = Stack2()
    for i in range(10):
        stack.push(i)
    print(stack.peek())
    stack.data.print_list()
    value = stack.pop()
    while (value != None):
        print(value)
        value = stack.pop()
    print(stack.pop())
    print(stack.pop())
    print(stack.pop())
    print(stack.pop())

        

    

File: c/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "linked_list.c"
#include <stdbool.h>

typedef struct Stack
{
    LinkedList *data;
    Node *top;
} Stack;

Stack *get_stack(void)
{
    Stack *stack = (Stack *)malloc(sizeof(Stack));
    LinkedList *lst = (LinkedList *)malloc(sizeof(LinkedList));
    lst->head = NULL;
    lst->tail = NULL;
    lst->length = 0;
    stack->data = lst;
    stack->top = NULL;
    return stack;
}

void push(Stack *stack, char *item)
{
    Node *node = (Node *)malloc(sizeof(Node));
    node->value = item;
    node->next = NULL;
    stack->top = node;
    append_node(stack->data, &node);
}

Node *pop(Stack *stack)
{
    int length = stack->data->length;
    if (length > 0)
    {
        Node *node = delete_node(stack->data, length-1);
        if (length == 1)
            stack->top = NULL;
        else
            stack->top = stack->data->tail;
        return node;
    }
    return NULL;
}

Node *peek(Stack *stack)
{
    return stack->top;
}

bool is_empty(Stack *stack)
{
    if (stack->data->length > 0)
        return false;
    return true;
}

void free_stack(Stack *stack)
{
    if (stack != NULL)
    {
        free_list(stack->data);
        free(stack);
    }
}

int main(void)
{
    Stack *stack = get_stack();
    push(stack, "Hello world");
    printf("Peek = %s\n", peek(stack)->value);
    print_linked_list(stack->data);
    push(stack, "Hi");
    printf("Peek = %s\n", peek(stack)->value);
    print_linked_list(stack->data);
    push(stack, "Man");
    printf("Peek = %s\n", peek(stack)->value);
    print_linked_list(stack->data);    
    pop(stack);
    printf("Peek = %s\n", peek(stack)->value);
    print_linked_list(stack->data);
    pop(stack);
    printf("Peek = %s\n", peek(stack)->value);
    print_linked_list(stack->data);
    pop(stack);
    print_linked_list(stack->data);
    pop(stack);
    pop(stack);
    pop(stack);
    pop(stack);
    free_stack(stack);
}